src/scales.js

```diff
@@ -144,3 +175,32 @@ export function isCollapsed(scale) {
   }
   return true;
 }
+
+// Mutates channel.value!
+function coerceType(channels, options, coerce, type) {
+  for (const c of channels) c.value = coerceArray(c.value, coerce, type);
+  return {...options, domain: coerceArray(options.domain, coerce, type)};
+}
+
+function coerceArray(array, coerce, type = Array) {
+  if (array !== undefined) return type.from(array, coerce);
+}
+
+// Unlike Mark’s number, here we want to convert null and undefined to NaN,
+// since the result will be stored in a Float64Array and we don’t want null to
+// be coerced to zero.
+function coerceNumber(x) {
+  return x == null ? NaN : +x;
+}
+
+// When coercing strings to dates, we only want to allow the ISO 8601 format
+// since the built-in string parsing of the Date constructor varies across
+// browsers. (In the future, this could be made more liberal if desired, though
+// it is still generally preferable to do date parsing yourself explicitly,
+// rather than rely on Plot.) Any non-string values are coerced to number first
+// and treated as milliseconds since UNIX epoch.
+function coerceDate(x) {
+  return x instanceof Date ? x
+    : typeof x === "string" ? isoParse(x)
+    : new Date(x == null ? NaN : +x);
```

Comment:
Returning a Date instance for undefined or NaN input feels wrong here, especially because our defined helper considers invalid dates to be defined. We should probably return undefined instead of an invalid date.